### PR TITLE
[query] Add partition interval IR support

### PR DESCRIPTION
**Pull request by tpoterba**

### CODE REVIEW

File: hail/src/test/scala/is/hail/utils/RowIntervalSuite.scala [hunk 1]
```diff
@@ -1,28 +1,55 @@
...
+
+    assertContains(Interval(Row(0, 1), Row(1, 2, 4), true, true), Row(0, 1, 5))
+    assertContains(Interval(Row(0, 1), Row(1, 2, 4), false, true), Row(0, 1, 5), shouldContain = false)
+    assertContains(Interval(Row(0, 1), Row(0, 1, 4), true, true), Row(0, 1, 4))
+    assertContains(Interval(Row(0, 1), Row(0, 1, 4), true, false), Row(0, 1, 4), shouldContain = false)
+
+    assertContains(Interval(Row(0, 1), Row(1, 2, 4), true, true), Row(0, 1, 5))
+    assertContains(Interval(Row(0, 1), Row(1, 2, 4), false, true), Row(0, 1, 5), shouldContain = false)
+    assertContains(Interval(Row(0, 1), Row(0, 1, 4), true, true), Row(0, 1, 4))
+    assertContains(Interval(Row(0, 1), Row(0, 1, 4), true, false), Row(0, 1, 4), shouldContain = false)
```
Comment (patrick-schultz) [inline hail/src/test/scala/is/hail/utils/RowIntervalSuite.scala:52]: I'd like to have some tests cover zero-length endpoints too. Maybe
```
assertContains(Interval(Row(), Row(1, 2, 4), true, true), Row(1, 2, 4))
assertContains(Interval(Row(), Row(1, 2, 4), true, false), Row(1, 2, 4), shouldContain = false)
assertContains(Interval(Row(1, 2, 4), Row(), true, true), Row(1, 2, 4))
assertContains(Interval(Row(1, 2, 4), Row(), false, true), Row(1, 2, 4), shouldContain = false)
```

File: hail/src/main/scala/is/hail/expr/ir/functions/IntervalFunctions.scala [hunk 2]
```diff
@@ -148,6 +148,49 @@ object IntervalFunctions extends RegistryFunctions {
...
     }
+
+
+    val endpointT = TTuple(tv("T"), TInt32)
+    registerIR2("partitionIntervalContains",
+      TStruct("left" -> endpointT, "right" -> endpointT, "includesLeft" -> TBoolean, "includesRight" -> TBoolean),
+      tv("T"), TBoolean) {
+      case (_, interval, point) =>
+
+        def compareStructs(left: IR, right: IR, truncatedValue: Int): IR = {
```
[the remark this answers is not in the document]
Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/functions/IntervalFunctions.scala:159]: oh, can go away. It is a remnant of an incorrect implementation

File: hail/src/main/scala/is/hail/expr/ir/functions/IntervalFunctions.scala [hunk 3]
```diff
@@ -148,6 +148,49 @@ object IntervalFunctions extends RegistryFunctions {
...
+          }
+        }
+
+        bindIRs(point, GetField(interval, "left"), GetField(interval, "right")) { case Seq(point, l, r) =>
+
+
+          val gtEqLeft = bindIR(compareStructs(l, point, 0)) { lc =>
+            If(lc > 0,
+              False(),
+              (lc < 0) || GetField(interval, "includesLeft"))
```
[the remark this answers is not in the document]
Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/ir/functions/IntervalFunctions.scala:183]: ah, that is better.

File: hail/src/test/scala/is/hail/utils/RowIntervalSuite.scala (hunk 1 above)
Comment (tpoterba) [reply hail/src/test/scala/is/hail/utils/RowIntervalSuite.scala:52]: oh, good call

**Comment by tpoterba** [comment]
good comments, thanks! addressed.

**Pull request merged**